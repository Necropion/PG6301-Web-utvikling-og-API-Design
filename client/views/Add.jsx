import Header from "../components/view/Header";
import Footer from "../components/view/Footer";
import { useNavigate } from "react-router-dom";
import { useContext, useEffect, useState } from "react";
import useLoginContext from "../hooks/useLoginContext";
import Navigation from "../components/view/Navigation";
import ProfileCard from "../components/view/ProfileCard";

const Add = () => {
  const addNav = useNavigate();

  const { username, userID, googleName, googleEmail, loadUser } =
    useContext(useLoginContext);

  const [title, setTitle] = useState("");
  const [description, setDescription] = useState("");
  const [inputError, setInputError] = useState("");

  const checkAvailability = async (e) => {
    e.preventDefault();

    const searchChat = await fetch("/api/chat/availibility/" + title);

    const foundChat = await searchChat.json();

    if (!searchChat.ok) {
      console.log("Something went wrong when trying to create a chat");
    }

    if (searchChat.ok) {
      if (foundChat.availibility === "available") {
        await handleAdd();
        setInputError("");
      }

      if (foundChat.availibility === "unavailable") {
        setInputError(
          "Sorry but the name for that chat is already taken, try a different one",
        );
      }
    }
  };

  const handleAdd = async () => {
    if (username) {
      const newChat = {
        title,
        description,
        user: username,
        userID: userID.toString(),
      };

      try {
        const response = await fetch("/api/chat", {
          method: "POST",
          body: JSON.stringify(newChat),
          headers: {
            "Content-Type": "application/json",
          },
        });

        if (!response.ok) {
          console.log(
            `Something went wrong while posting new task: ${response.status} ${response.statusText}`,
          );
        }

        if (response.ok) {
          console.log("Task added successfully!");
          addNav("/my-tasks");
        }
      } catch (error) {
        console.log(error);
      }
    }

    if (googleName) {
      const newChat = {
        title,
        description,
        user: googleName,
        userID: googleEmail.toString(),
      };

      try {
        const response = await fetch("/api/chat", {
          method: "POST",
          body: JSON.stringify(newChat),
          headers: {
            "Content-Type": "application/json",
          },
        });

        if (!response.ok) {
          console.log(
            `Something went wrong while posting new task: ${response.status} ${response.statusText}`,
          );
        }

        if (response.ok) {
          console.log("Task added successfully!");
          addNav("/my-tasks");
        }
      } catch (error) {
        console.log(error);
      }
    }
  };

  useEffect(() => {
    loadUser();
  }, []);

  return (
    <div className="application">
      <div className="header">
        <Header />
      </div>
      <div className="main">
        <div className="view">
          <Navigation />
          <div className="viewOutput">
            <div className="addOutput">
              <div></div>
              <form className="addTask" onSubmit={checkAvailability}>
                <h3>Add Chat</h3>
                <p>{inputError}</p>
                <div className="inputDiv">
                  <div className="title">
                    <p>Title</p>
                    <input
                      type="text"
                      placeholder="title"
                      value={title}
                      onChange={(e) => setTitle(e.target.value)}
                      required
                    />
                  </div>
                  <div className="description">
                    <p>Description</p>
                    <input
                      className="descriptionInput"
                      type="text"
                      placeholder="description"
                      value={description}
                      onChange={(e) => setDescription(e.target.value)}
                      required
                    />
                  </div>
                  <div className="status">
                    <p></p>
                    <div></div>
                  </div>
                </div>
                <button className="viewBtn" value="addTask">
                  Add
                </button>
              </form>
              <ProfileCard />
            </div>
          </div>
        </div>
      </div>
      <div className="footer">
        <Footer />
      </div>
    </div>
  );
};

export default Add;
