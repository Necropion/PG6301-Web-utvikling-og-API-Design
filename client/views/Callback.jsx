import Header from "../components/view/Header";
import Footer from "../components/view/Footer";
import { useContext, useEffect, useState } from "react";
import useLoginContext from "../hooks/useLoginContext";
import { useNavigate } from "react-router-dom";

const Callback = () => {
  const profileNav = useNavigate();
  const [debug, setDebug] = useState();
  const { loadUser } = useContext(useLoginContext);

  const handleClick = async (e) => {
    e.preventDefault();

    try {
      if (e.target.value === "home") {
        profileNav("/home");
      }

      if (e.target.value === "profile") {
        profileNav("/profile");
      }

      if (e.target.value === "logout") {
        const response = await fetch("/api/login", {
          method: "DELETE",
        });

        if (!response.ok) {
          throw new Error(
            `Something wnt wrong ${response.status} ${response.statusText}`,
          );
        }

        if (response.ok) {
          profileNav("/");
        }
      }
    } catch (error) {
      console.log(error);
    }
  };

  const handleCallback = async () => {
    try {
      const hashObject = Object.fromEntries(
        new URLSearchParams(window.location.hash.substring(1)),
      );
      const { access_token } = hashObject;

      const response = await fetch("/api/google", {
        method: "POST",
        body: JSON.stringify({ access_token }),
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        const errorMsg = `Failed POST '/api/google': ${response.status} ${response.statusText}`;
        setDebug(errorMsg);
      }

      if (response.ok) {
        await fetchGoogle();
      }
    } catch (error) {
      console.log(error);
    }
  };

  const fetchGoogle = async () => {
    try {
      const response = await fetch("/api/google");

      const userInfo = await response.json();

      window.sessionStorage.setItem("googleName", userInfo.name);
      window.sessionStorage.setItem("googleEmail", userInfo.email);
      window.sessionStorage.setItem("googleImg", userInfo.picture);
      await checkGoogleUser(userInfo.name, userInfo.picture, userInfo.email);
      profileNav("/home");
    } catch (error) {
      console.log("Error while trying to fetch googleUser: " + error);
    }
  };

  const checkGoogleUser = async (name, img, email) => {
    try {
      const checkUser = await fetch("/api/google/user/" + email);

      const foundUser = await checkUser.json();

      if (checkUser.ok) {
        if (foundUser.search === "failed") {
          await createGoogleUser(name, img, email);
        }
      }

      if (!checkUser.ok) {
        console.log("Something went wrong checking the user ");
      }
    } catch (error) {
      console.log("Error while checking for google user " + error);
    }
  };

  const createGoogleUser = async (name, img, email) => {
    try {
      const googleCredentials = { name, img, email };

      const createUser = await fetch("/api/google/user", {
        method: "POST",
        body: JSON.stringify(googleCredentials),
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (createUser.ok) {
        console.log("Google user added!");
      }
      if (!createUser.ok) {
        console.log("Something went wrong creating the user ");
      }
    } catch (error) {
      console.log("Error while trying to create google user " + error);
    }
  };

  useEffect(() => {
    handleCallback();
  });

  return (
    <div className="application">
      <div className="header">
        <Header />
      </div>
      <div className="main">
        <div className="view">
          <div className="viewDiv">
            <h1 className="viewHeader">Please Wait...</h1>
            <button className="viewBtn" value="home" onClick={handleClick}>
              Home
            </button>
            <button className="viewBtn" value="profile" onClick={handleClick}>
              Profile
            </button>
            <button className="viewBtn" value="logout" onClick={handleClick}>
              Sign out
            </button>
          </div>
          <div className="viewOutput">
            <pre>{debug}</pre>
          </div>
        </div>
      </div>
      <div className="footer">
        <Footer />
      </div>
    </div>
  );
};

export default Callback;
