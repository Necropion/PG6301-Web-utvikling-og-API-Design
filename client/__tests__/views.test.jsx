import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import Add from "../views/Add";
import UseLoginContext from "../hooks/useLoginContext";
import Callback from "../views/Callback";

describe("View tests", () => {
  it("Renders Add View", () => {
    const inputTitle = "Games";
    const inputDescription = "adfg";

    const component = renderer.create(
      <UseLoginContext.Provider
        value={{
          inputTitle,
          inputDescription,
          username: "Henry",
          userID: "655e8c432fff7d8ef2a281fc",
          googleName: "Jokubas Gedminas",
          googleEmail: "jokas3@gmail.com",
        }}
      >
        <MemoryRouter>
          <Add />
        </MemoryRouter>
      </UseLoginContext.Provider>,
    );
    expect(component).toMatchSnapshot();
  });

  it("Renders Callback", () => {
    const component = renderer.create(
      <MemoryRouter>
        <Callback />
      </MemoryRouter>,
    );
    expect(component).toMatchSnapshot();
  });
});
