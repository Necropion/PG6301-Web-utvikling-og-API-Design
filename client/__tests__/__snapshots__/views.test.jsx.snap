// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View tests Renders Add View 1`] = `
<div
  className="application"
>
  <div
    className="header"
  >
    <h1
      className="headerTitle"
    >
      Simple Application
    </h1>
  </div>
  <div
    className="main"
  >
    <div
      className="view"
    >
      <div
        className="viewDiv"
      >
        <h1
          className="viewHeader"
        >
          Welcome 
          Jokubas Gedminas
        </h1>
        <button
          className="viewBtn"
          onClick={[Function]}
          value="home"
        >
          Home
        </button>
        <button
          className="viewBtn"
          onClick={[Function]}
          value="my-tasks"
        >
          My Chats
        </button>
        <button
          className="viewBtn"
          onClick={[Function]}
          value="profile"
        >
          Profile
        </button>
        <button
          className="viewBtn"
          onClick={[Function]}
          value="logout"
        >
          Sign out
        </button>
      </div>
      <div
        className="viewOutput"
      >
        <div
          className="addOutput"
        >
          <div />
          <form
            className="addTask"
            onSubmit={[Function]}
          >
            <h3>
              Add Chat
            </h3>
            <p />
            <div
              className="inputDiv"
            >
              <div
                className="title"
              >
                <p>
                  Title
                </p>
                <input
                  onChange={[Function]}
                  placeholder="title"
                  required={true}
                  type="text"
                  value=""
                />
              </div>
              <div
                className="description"
              >
                <p>
                  Description
                </p>
                <input
                  className="descriptionInput"
                  onChange={[Function]}
                  placeholder="description"
                  required={true}
                  type="text"
                  value=""
                />
              </div>
              <div
                className="status"
              >
                <p />
                <div />
              </div>
            </div>
            <button
              className="viewBtn"
              value="addTask"
            >
              Add
            </button>
          </form>
          <div
            className="profileDiv"
          >
            <img
              alt="No Image"
              placeholder="No Image"
              src="https://static.vecteezy.com/system/resources/thumbnails/020/765/399/small/default-profile-account-unknown-icon-black-silhouette-free-vector.jpg"
            />
            <h3>
              User
            </h3>
            <p>
              Username: Henry
            </p>
            <p>
              UserID: 655e8c432fff7d8ef2a281fc
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="footer"
  >
    <div>
      Made by Jokubas Gedminas
    </div>
  </div>
</div>
`;

exports[`View tests Renders Callback 1`] = `
<div
  className="application"
>
  <div
    className="header"
  >
    <h1
      className="headerTitle"
    >
      Simple Application
    </h1>
  </div>
  <div
    className="main"
  >
    <div
      className="view"
    >
      <div
        className="viewDiv"
      >
        <h1
          className="viewHeader"
        >
          Please Wait...
        </h1>
        <button
          className="viewBtn"
          onClick={[Function]}
          value="home"
        >
          Home
        </button>
        <button
          className="viewBtn"
          onClick={[Function]}
          value="profile"
        >
          Profile
        </button>
        <button
          className="viewBtn"
          onClick={[Function]}
          value="logout"
        >
          Sign out
        </button>
      </div>
      <div
        className="viewOutput"
      >
        <pre />
      </div>
    </div>
  </div>
  <div
    className="footer"
  >
    <div>
      Made by Jokubas Gedminas
    </div>
  </div>
</div>
`;
